source/extensions/filters/network/mysql_proxy/mysql_terminal_filter.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/api/api.h"
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/tcp/conn_pool.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+
+#include "extensions/filters/network/mysql_proxy/mysql_codec.h"
+#include "extensions/filters/network/mysql_proxy/mysql_codec_clogin.h"
+#include "extensions/filters/network/mysql_proxy/mysql_codec_clogin_resp.h"
+#include "extensions/filters/network/mysql_proxy/mysql_codec_command.h"
+#include "extensions/filters/network/mysql_proxy/mysql_codec_greeting.h"
+#include "extensions/filters/network/mysql_proxy/mysql_codec_switch_resp.h"
+#include "extensions/filters/network/mysql_proxy/mysql_decoder.h"
+#include "extensions/filters/network/mysql_proxy/mysql_session.h"
+#include "extensions/filters/network/mysql_proxy/route.h"
+#include "extensions/filters/network/mysql_proxy/stats.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace MySQLProxy {
+
+/**
+ * Implementation of MySQL proxy filter.
+ */
+class MySQLTerminalFilter : public Tcp::ConnectionPool::Callbacks,
+                            public Network::ReadFilter,
+                            public Logger::Loggable<Logger::Id::filter> {
+
+  friend class MySQLFilterTest;
+  friend class MySQLTerminalFitlerTest;
+
+public:
+  MySQLTerminalFilter(MySQLFilterConfigSharedPtr config, RouterSharedPtr router,
```

Comment:
It seems that there are two separate classes handling MySQL traffic now: `MySQLTerminalFilter` and `MySQLFilter`. There is lots of overlap between them. I believe that differences could be handled via virtual methods or at least by defining terminal filter as descendant from `MySQLFilter` and reusing `MySQLFilter`'s methods and calling extra logic to support routing part. I am afraid that any bug fixes/improvements will have to be done in two classes.

Reply:
Well, I think my current implementation could be more clear after your mention😊.